### PR TITLE
MBS-8580: (release, position) conflict in "Reorder mediums" edit

### PR DESCRIPTION
**Pull request by mwiencek**
Before accepting the edit, check if there exist any mediums which are already in positions we want to move into (unless those mediums themselves are being moved out of those positions). If so, throw a `FailedDependency` exception to cause the edit to fail and to have ModBot leave a note.

### CODE REVIEW

**Comment by mwiencek** [comment]
I think we'll need to upgrade Jira to get the links to work, not seeing a way to hook things together in 4.0

**Comment by gentlecat** [comment]
Ah, right. I forgot about links to Jira.

File: lib/MusicBrainz/Server/Edit/Release/ReorderMediums.pm [hunk 1]
```diff
@@ -103,10 +104,28 @@ sub build_display_data {
...
+              FROM medium
+             WHERE release = ?
+               AND position = any(?)
+EOSQL
+
+    for my $row (@$possible_conflicts) {
+        unless (exists $medium_positions{$row->{id}}) {
+            MusicBrainz::Server::Edit::Exceptions::FailedDependency->throw(
+                'Can’t move a medium into position ' . $row->{position} . ', where one already exists.'
+            );
```
Comment (gentlecat) [inline lib/MusicBrainz/Server/Edit/Release/ReorderMediums.pm]: What happens after this exception is thrown (or in general when edit fails)? I assume ModBot needs to get a note about it somehow.

File: lib/MusicBrainz/Server/Edit/Release/ReorderMediums.pm [hunk 2]
```diff
@@ -1,4 +1,5 @@
 package MusicBrainz::Server::Edit::Release::ReorderMediums;
+use utf8;
```
Comment (gentlecat) [inline lib/MusicBrainz/Server/Edit/Release/ReorderMediums.pm]: Why is that needed?

**Comment by gentlecat** [comment]
LGTM

File: lib/MusicBrainz/Server/Edit/Release/ReorderMediums.pm (hunk 1 above)
Comment (mwiencek) [inline lib/MusicBrainz/Server/Edit/Release/ReorderMediums.pm]: Right, ModBot will catch the exception and do what's necessary. The basic flow of operations is this:

1. admin/CheckVotes.pl ("ModBot") is run hourly on astro, via admin/cron/hourly.sh
2. `MusicBrainz::Server::EditQueue::process_edits` is called from CheckVotes.pl
3. `process_edits` will loop over open/expired edits, and try to accept relevant ones [here](https://github.com/mwiencek/musicbrainz-server/blob/1b01b1f/lib/MusicBrainz/Server/EditQueue.pm#L160).
4. That'll invoke the `accept` method on the edit, the one I'm changing here, [within a try/catch block](https://github.com/mwiencek/musicbrainz-server/blob/1b01b1f/lib/MusicBrainz/Server/Data/Edit.pm#L682).
5. The `FailedDependency` exception [is caught](https://github.com/mwiencek/musicbrainz-server/blob/1b01b1f/lib/MusicBrainz/Server/Data/Edit.pm#L687), and a note containing the error message is left by the ModBot editor account.

File: lib/MusicBrainz/Server/Edit/Release/ReorderMediums.pm (hunk 2 above)
Comment (mwiencek) [inline lib/MusicBrainz/Server/Edit/Release/ReorderMediums.pm]: It only needs adding when you use UTF-8 chars in the source code, and I used some fancy quotes in the exception message ("Can’t")

File: lib/MusicBrainz/Server/Edit/Release/ReorderMediums.pm (hunk 1 above)
Comment (gentlecat) [inline lib/MusicBrainz/Server/Edit/Release/ReorderMediums.pm]: And what happens to the edit? Is it automatically rejected when `FailedDependency` exception is caught?

File: lib/MusicBrainz/Server/Edit/Release/ReorderMediums.pm (hunk 2 above)
Comment (gentlecat) [inline lib/MusicBrainz/Server/Edit/Release/ReorderMediums.pm]: Ah, OK. So kind of like `# -*- coding: utf-8 -*-` in Python 2.

File: lib/MusicBrainz/Server/Edit/Release/ReorderMediums.pm (hunk 1 above)
Comment (mwiencek) [inline lib/MusicBrainz/Server/Edit/Release/ReorderMediums.pm]: Yes, the result of `_do_accept` (in this case `$STATUS_FAILEDDEP`) is immediately passed to `_close`, which updates the status and close time of the edit no matter what. So edits only get "stuck" when they throw an exception that isn't handled by `_do_accept`.

Comment (gentlecat) [inline lib/MusicBrainz/Server/Edit/Release/ReorderMediums.pm]: Understood. :+1:

File: lib/MusicBrainz/Server/Edit/Release/ReorderMediums.pm (hunk 2 above)
Comment (mwiencek) [inline lib/MusicBrainz/Server/Edit/Release/ReorderMediums.pm]: Yeah, I forget the details of that but looks about the same. :)

Comment (lalinsky) [inline lib/MusicBrainz/Server/Edit/Release/ReorderMediums.pm]: If I remember correctly, this is more like `from __future__ import unicode_literals` combined with `# -*- coding: utf-8 -*-`.
